pkg/apis/serving/v1alpha2/kfservice_types.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2019 kubeflow.org.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"github.com/knative/pkg/apis"
+	duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// KFServiceSpec defines the desired state of KFService
+type KFServiceSpec struct {
+	// Default defines the configuration of KFService
+	Default ServicesSpec `json:"default"`
+	// Canary defines an alternate KFService configuration to route a percentage of traffic.
+	Canary *ServicesSpec `json:"canary,omitempty"`
+	// Traffic split percentage between default and canary KFService
+	CanaryTrafficPercent int `json:"canaryTrafficPercent,omitempty"`
+}
+
+type ServicesSpec struct {
+	// Optional prepost processing specification
+	Prepostprocess *PrepostprocessSpec `json:"prepostprocess,omitempty"`
+	// Predictor for a given model
+	Predict ModelSpec `json:"predict"`
+	// Optional Explain specification to add a model explainer next to the chosen predictor.
+	Explain *ExplainSpec `json:"explain,omitempty"`
+	// Default deployment specification
+	Deployment *DeploymentSpec `json:"deployment,omitempty"`
+	// Service account name
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+}
+
+type DeploymentSpec struct {
+	// Minimum number of replicas, pods won't scale down to 0 in case of no traffic
+	MinReplicas int `json:"minReplicas,omitempty"`
+	// This is the up bound for autoscaler to scale to
+	MaxReplicas int `json:"maxReplicas,omitempty"`
+	// Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// ModelSpec defines the configuration to route traffic to a predictor.
+type ModelSpec struct {
+	// The following fields follow a "1-of" semantic. Users must specify exactly one spec.
+	Custom     *CustomSpec     `json:"custom,omitempty"`
+	Tensorflow *TensorflowSpec `json:"tensorflow,omitempty"`
+	TensorRT   *TensorRTSpec   `json:"tensorrt,omitempty"`
+	XGBoost    *XGBoostSpec    `json:"xgboost,omitempty"`
+	SKLearn    *SKLearnSpec    `json:"sklearn,omitempty"`
+	PyTorch    *PyTorchSpec    `json:"pytorch,omitempty"`
+}
+
+type PrepostprocessSpec struct {
```

Comment:
I feel a little bit that Preprocessor and PostProcessor and a little unwieldy as terms.

I wonder if there's another way we could think of them. I kind of like Seldon's "Transformer" concept. Perhaps we could take a list of transformers and part of the transformer spec could be `type={"before", "after", "both" (default)}. We'd need some semantic for what order they're executed in, but I think FIFO is probably clean enough.

What do you think?

Reply:
+1 to transformer..

Reply:
Also thinking a bit ahead and a graph sitting above kfserving each node if really doing a "transform". Might be great to allow "transform" to do whole sequence of pre->predict->post. That could be part of data plane discussions.

Reply:
PreProcessor and PostProcessor are well accepted in industry - and most of the users would understand what they mean in context of models.

Reply:
I am reading from here https://www.tensorflow.org/tfx/transform/get_started, we can name it transformer on the spec and user implements preprocess_fn and postprocess_fn according to our kfserver interface.

Reply:
I like this a lot!